File: examples/contrib/word_square.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Text.RegularExpressions;
using Google.OrTools.ConstraintSolver;

public class WordSquare
{
    /**
     *
     * Word square.
     *
     * From http://en.wikipedia.org/wiki/Word_square
     * """
     * A word square is a special case of acrostic. It consists of a set of words,
     * all having the same number of letters as the total number of words (the
     * 'order' of the square); when the words are written out in a square grid
     * horizontally, the same set of words can be read vertically.
     * """
     *
     * See http://www.hakank.org/or-tools/word_square.py
     *
     */
    private static void Solve(String[] words, int word_len, int num_answers)
    {
        Solver solver = new Solver("WordSquare");

        int num_words = words.Length;
        Console.WriteLine("num_words: " + num_words);
        int n = word_len;
        IEnumerable<int> WORDLEN = Enumerable.Range(0, word_len);

        //
        //  convert a character to integer
        //

        String alpha = "abcdefghijklmnopqrstuvwxyz";
        Hashtable d = new Hashtable();
        Hashtable rev = new Hashtable();
        int count = 1;
        for (int a = 0; a < alpha.Length; a++)
        {
            d[alpha[a]] = count;
            rev[count] = a;
            count++;
        }

        int num_letters = alpha.Length;

        //
        // Decision variables
        //
        IntVar[,] A = solver.MakeIntVarMatrix(num_words, word_len, 0, num_letters, "A");
        IntVar[] A_flat = A.Flatten();
        IntVar[] E = solver.MakeIntVarArray(n, 0, num_words, "E");

        //
        // Constraints
        //
        solver.Add(E.AllDifferent());

        // copy the words to a matrix
        for (int i = 0; i < num_words; i++)
        {
            char[] s = words[i].ToArray();
            foreach (int j in WORDLEN)
            {
                int t = (int)d[s[j]];
                solver.Add(A[i, j] == t);
            }
        }

        foreach (int i in WORDLEN)
        {
            foreach (int j in WORDLEN)
            {
                solver.Add(A_flat.Element(E[i] * word_len + j) == A_flat.Element(E[j] * word_len + i));
            }
        }

        //
        // Search
        //
        DecisionBuilder db =
            solver.MakePhase(E.Concat(A_flat).ToArray(), Solver.CHOOSE_FIRST_UNBOUND, Solver.ASSIGN_MIN_VALUE);

        solver.NewSearch(db);

        int num_sols = 0;
        while (solver.NextSolution())
        {
            num_sols++;
            for (int i = 0; i < n; i++)
            {
                Console.WriteLine(words[E[i].Value()] + " ");
            }
            Console.WriteLine();

            if (num_answers > 0 && num_sols >= num_answers)
            {
                break;
            }
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    /*
     *
     * Read the words from a word list with a specific word length.
     *
     */
    public static String[] ReadWords(String word_list, int word_len)
    {
        Console.WriteLine("ReadWords {0} {1}", word_list, word_len);
        List<String> all_words = new List<String>();

        TextReader inr = new StreamReader(word_list);
        String str;
        int count = 0;
        Hashtable d = new Hashtable();
        while ((str = inr.ReadLine()) != null)
        {
            str = str.Trim().ToLower();
            // skip weird words
            if (Regex.Match(str, @"[^a-z]").Success || d.Contains(str) || str.Length == 0 || str.Length != word_len)
            {
                continue;
            }

            d[str] = 1;
            all_words.Add(str);
            count++;

        } // end while

        inr.Close();

        return all_words.ToArray();
    }

    public static void Main(String[] args)
    {
        String word_list = "/usr/share/dict/words";
        int word_len = 4;
        int num_answers = 20;

        if (args.Length > 0)
        {
            word_list = args[0];
        }

        if (args.Length > 1)
        {
            word_len = Convert.ToInt32(args[1]);
        }

        if (args.Length > 2)
        {
            num_answers = Convert.ToInt32(args[2]);
        }

        String[] words = ReadWords(word_list, word_len);

        Solve(words, word_len, num_answers);
    }
}

File: examples/contrib/word_square.py
#!/usr/bin/env python3
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Word square in Google CP Solver.

From http://en.wikipedia.org/wiki/Word_square ''' A word square is a special
case of acrostic. It consists of a set of words, all having the same number of
letters as the total number of words (the 'order' of the square); when the words
are written out in a square grid horizontally, the same set of words can be read
vertically. '''

Compare with the following models:
* MiniZinc: http://www.hakank.org/minizinc/word_square.mzn
* Comet   : http://www.hakank.org/comet/word_square.co
* Choco   : http://www.hakank.org/choco/WordSquare.java
* Gecode  : http://www.hakank.org/gecode/word_square.cpp
* Gecode  : http://www.hakank.org/gecode/word_square2.cpp
* JaCoP   : http://www.hakank.org/JaCoP/WordSquare.java
* Zinc: http://hakank.org/minizinc/word_square.zinc

This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

import re
import sys

from ortools.constraint_solver.python import constraint_solver as cp


def main(words, word_len, num_answers=20):
    # Create the solver.
    solver = cp.Solver("Problem")

    #
    # data
    #
    num_words = len(words)
    n = word_len
    d, rev = get_dict()

    #
    # declare variables
    #
    A = {}
    for i in range(num_words):
        for j in range(word_len):
            A[(i, j)] = solver.new_int_var(0, 29, "A(%i,%i)" % (i, j))

    A_flat = [A[(i, j)] for i in range(num_words) for j in range(word_len)]

    E = [solver.new_int_var(0, num_words, "E%i" % i) for i in range(n)]

    #
    # constraints
    #
    solver.add_all_different(E)

    # copy the words to a Matrix
    for I in range(num_words):
        for J in range(word_len):
            solver.add(A[(I, J)] == d[words[I][J]])

    for i in range(word_len):
        for j in range(word_len):
            # This is what I would like to do:
            # solver.add(A[(E[i],j)] == A[(E[j],i)])

            # We must use Element explicitly
            solver.add(
                solver.element(A_flat, E[i] * word_len + j)
                == solver.element(A_flat, E[j] * word_len + i)
            )

    #
    # solution and search
    #
    solution = solver.assignment()
    solution.add(E)

    # db: DecisionBuilder
    db = solver.phase(
        E + A_flat,
        cp.IntVarStrategy.CHOOSE_FIRST_UNBOUND,
        cp.IntValueStrategy.ASSIGN_MIN_VALUE,
    )

    solver.new_search(db)
    num_solutions = 0
    while solver.next_solution():
        # print E
        print_solution(E, words)
        num_solutions += 1
        if num_solutions > num_answers:
            break

    solver.end_search()

    print()
    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


#
# convert a character to integer
#
def get_dict():
    alpha = "abcdefghijklmnopqrstuvwxyzåäö"
    d = {}
    rev = {}
    count = 1
    for a in alpha:
        d[a] = count
        rev[count] = a
        count += 1
    return d, rev


def print_solution(E, words):
    # print E
    for e in E:
        print(words[e.value()])
    print()


def read_words(word_list, word_len, limit):
    dict = {}
    all_words = []
    count = 0
    words = open(word_list).readlines()
    for w in words:
        w = w.strip().lower()
        # if len(w) == word_len and not dict.has_key(w) and not re.search("[^a-zåäö]",w) and count < limit:
        # Later note: The limit is not needed anymore with Mistral
        if len(w) == word_len and w not in dict and not re.search("[^a-zåäö]", w):
            dict[w] = 1
            all_words.append(w)
            count += 1
    return all_words


word_dict = "examples/contrib/words_list.txt"
word_len = 4
limit = 1000000
num_answers = 5

if __name__ == "__main__":
    if len(sys.argv) > 1:
        word_dict = sys.argv[1]
    if len(sys.argv) > 2:
        word_len = int(sys.argv[2])
    if len(sys.argv) > 3:
        limit = int(sys.argv[3])
    if len(sys.argv) > 4:
        num_answers = int(sys.argv[4])

    # Note: I have to use a limit, otherwise it seg faults
    words = read_words(word_dict, word_len, limit)
    print(f"It was {len(words)} words")
    main(words, word_len, num_answers)
